[IN REVIEW] add a wrapper for (hyper)elastic materials from the muesli library

henrij22 opened the pull request:
There was some wierd behaviour spotted in the python test for Muesli materials using Vanishing Stress. 
To reproduce this, you can change the line
https://github.com/ikarus-project/ikarus/blob/89bd438b17b184327869f7898bd1e42b151e6b5f/ikarus/python/test/testmaterials.py#L191
from 
```py
checkMaterial(mmoon.asPlaneStress(), strain, rCGT, 3)
``` 
to 
```py
checkMaterial(mmoon.asPlaneStress(), strain, gtT, 3)
``` 

which then uses the the same strain measure but with GreenLagrange tag. 
This behviour is not present in our NeoHooke Implementation. 
This might come from the fact that the strain tensor has to be converted to deformation gradient bevore its passed to Muesli

henrij22 committed:
try

works
tests + interface
add constructors
unique ptr + more
add python bindings
update CI workflows and improve Muesli material functions
python bindings

fix python?

python should work now?

henrij22 committed:
renaming + workflow fixes

henrij22 committed:
tidy up + reformat

henrij22 committed:
minor things

henrij22 committed:
documentation + changelog

henrij22 committed:
fomat

henrij22 committed:
refactor cmake flags for autodiff and muesli libraries

add muesli library citation and fix typos in documentation
Fix xmake + code review (I)
Minor fixes

add tests to testmaterials
add header guards for muesli

tarun-mitruka commented on CHANGELOG.md:
```diff
@@ -56,6 +56,7 @@ SPDX-License-Identifier: LGPL-3.0-or-later
 - Add an About Ikarus page in the documentation ([#291](https://github.com/ikarus-project/ikarus/pull/291))
 - Add new class `Vtk::Writer`, which implements some convenience methods over the existing `dune-vtk` module ([#309](https://github.com/ikarus-project/ikarus/pull/309))
 - Add `VanishingStrain` material (useful for example for plane strain case), also refactor the constructor of `LinearElastic` to take any linear material law ([#317](https://github.com/ikarus-project/ikarus/pull/317))
+- Adds wrapper classes to the muesli material for elastic and hyperelastic materials ([#337](https://github.com/ikarus-project/ikarus/pull/337))
```
Add wrapper classes to small strain and finite strain elastic material models from the Muesli library (...)

tarun-mitruka commented on cmake/modules/AddMuesliFlags.cmake:
```diff
@@ -0,0 +1,20 @@
+# add function to link against the muesli library
+function(add_dune_muesli_flags _targets)
+  if(HAVE_MUESLI)
+    foreach(_target ${_targets})
```
`message("Muesli added")` can be added here?

henrij22 replied:
I had that at one point to check if that works, but it was never printed, also the message i put in the Eigen file was never printed, so I am not sure how these cmake macros work

tarun-mitruka commented on ikarus/finiteelements/mechanics/materials/muesli/mueslifinite.hh:
```diff
@@ -0,0 +1,161 @@
+
+/**
+ * \file Muesli.hh
+ * \brief Implementation of the Muesli material model.
```
Implementation of the Muesli material model for finite strains.

tarun-mitruka commented on ikarus/finiteelements/mechanics/materials/muesli/mueslifinite.hh:
```diff
@@ -0,0 +1,161 @@
+
+/**
+ * \brief Wrapper class for finite strain materials from the muesli library. It can be templated with all materials
+ * derived from muesli::finiteStrainMaterial. It models the Ikarus material interface.
```
It adheres to the Ikarus material interface. See \ref Material (or similar)

tarun-mitruka commented on ikarus/finiteelements/mechanics/materials/muesli/mueslifinite.hh:
```diff
@@ -0,0 +1,161 @@
+ * derived from muesli::finiteStrainMaterial. It models the Ikarus material interface.
+ *
+ * \tparam FM muesli material model implementation
+ * \remark Please cite \cite portillo_muesli_2017 if you use any materials from the muesli library
```
Here I think `See \cite ...` would be enough. We can add a concept for Muesli-based materials. Then in `struct Material`, we can check if the underlying material `MI` is of Muesli type and then provide an `spdlog::info` to cite this. We can discuss this as it should then only log it once and not for every integration point.

tarun-mitruka replied:
I didn't notice the concept `MuesliMaterialImplementation`. Perhaps this can already be used

henrij22 replied:
For this to work I think we need a global variable to be set. 
I just had the idea to have global ikarus parameters, that are initialized through `init()` and can then be accessed throghout the code through a static singleton object.  But I think this has to be discussed in a seperate PR

tarun-mitruka replied:
Yes. I am not exactly sure how the implementation will look like atm but yes it can also be done separately in a different PR. Can you note it down in some issue?

tarun-mitruka commented on ikarus/finiteelements/mechanics/materials/muesli/mueslifinite.hh:
```diff
@@ -0,0 +1,161 @@
+  [[nodiscard]] constexpr static std::string nameImpl() noexcept { return "FiniteStrain: " + materialName<FM>(); }
+
+  /**
+   * \briefCConstructor for FiniteStrain muesli materials
```
fix spelling `CConstructor`

tarun-mitruka commented on ikarus/finiteelements/mechanics/materials/muesli/mueslifinite.hh:
```diff
@@ -0,0 +1,161 @@
+  MaterialParameters materialParametersImpl() const { return materialParameter_; }
+
+  /**
+   * \brief Computes the stored energy in the Neo-Hookean material model.
```
fix (everywhere in this file)

tarun-mitruka commented on ikarus/finiteelements/mechanics/materials/muesli/mueslihelpers.hh:
```diff
@@ -0,0 +1,124 @@
+
+/**
+ * \brief adds a tag to the muesli materialproperties that the material model should use regularized stretches (i.e.
+ * deviatoric principal stretchs), applicable to NeoHooke material model
```
typo `stretches`

tarun-mitruka commented on ikarus/finiteelements/mechanics/materials/muesli/mueslimaterials.hh:
```diff
@@ -0,0 +1,126 @@
+ *
+ * \tparam MPT the type of the Ikarus material parameters
+ * \param mpt the Ikarus material parameters
+ * \param useDeviatoricStretches tells the material to use deviatopric principal stretches
```
typo `deviatoric`

tarun-mitruka commented on ikarus/finiteelements/mechanics/materials/muesli/mueslihelpers.hh:
```diff
@@ -0,0 +1,124 @@
+ * \brief adds a tag to the muesli materialproperties that the material model should be incompressible, applicable to
+ * MooneyRivlin material model
+ */
+inline void addIncompressibleTag(MaterialProperties& mpm) { mpm.insert({"incompressible", 0}); }
```
Since these functions are particular only for specific materials, how about removing them and having only one function
```cpp
inline void addTag(MaterialProperties& mpm, const std::string& tagName = "compressible", bool tagValue = false){
mpm.insert({tagName, tagValue});
}
```
And then pass the specific arguments in the corresponding `make` functions in `mueslimaterials.hh`?

tarun-mitruka commented on ikarus/finiteelements/mechanics/materials/muesli/mueslimaterials.hh:
```diff
@@ -0,0 +1,126 @@
+/**
+ * \brief Constructs a muesli MooneyRivlin material model based on muesli::mooneyMaterial
+ *
+ * \param alpha alpha0, alpha1, alpha2 parameters in a std::array
```
may be we can add a remark that alpha0 is the bulk modulus. It might be confusing?

henrij22 replied:
Ah I didn't knew this, Thanks

henrij22 replied:
Should I rename all the variables according to Hyperelastic PR, maybe we can later then synchronise the Parameter structs as well?

tarun-mitruka replied:
Yes sure that's even better. But then you can perhaps wait until #333 is merged to avoid unnecessary iterations on the changes and mark this conversation as unresolved.

tarun-mitruka commented on ikarus/finiteelements/mechanics/materials/muesli/mueslifinite.hh:
```diff
@@ -0,0 +1,161 @@
+   */
+  bool assertMP() const { return mp_.get() != NULL; }
+
+  FiniteStrain(const FiniteStrain& other)
```
Is this needed? I don't recall the conclusion of our discussion on this?

henrij22 replied:
Yes its needed as MaterialModel doesn't have a default constructor.

henrij22 replied:
maybe we even need more, for example a copy assignement, etc...

tarun-mitruka commented on python/ikarus/materials/__init__.py:
```diff
@@ -0,0 +1,55 @@
+from dune.common.hashit import hashIt
+
+from ikarus.generator import MySimpleGenerator
+from dune.vtk import FormatTypes, DataTypes
```
`FormatTypes` can be removed?

tarun-mitruka commented on python/ikarus/materials/__init__.py:
```diff
@@ -0,0 +1,55 @@
+
+from ikarus.generator import MySimpleGenerator
+from dune.vtk import FormatTypes, DataTypes
+import dune.vtk
```
can be removed?

tarun-mitruka commented on python/ikarus/materials/__init__.py:
```diff
@@ -0,0 +1,55 @@
+from dune.vtk import FormatTypes, DataTypes
+import dune.vtk
+
+import types
```
also this?

tarun-mitruka commented on tests/src/testmaterial.cpp:
```diff
@@ -305,6 +303,18 @@ int main(int argc, char** argv) {
+  auto muesliNeoHooke = Materials::Muesli::makeNeoHooke(matPar, false);
+  t.subTest(testMaterial(muesliNeoHooke));
+
+  auto muesliNeoHookeReg = Materials::Muesli::makeNeoHooke(matPar, true);
```
May be we can add similar tests for other finite strain material models?

henrij22 replied:
Ah yes I wanted to do this, I was just unsure about good values for the material parameters

tarun-mitruka commented on tests/src/testmaterialswithmuesli.cpp:
```diff
@@ -0,0 +1,134 @@
+using Dune::TestSuite;
+
+template <StrainTags strainTag, typename MuesliMAT, typename IkarusMAT>
+auto testMaterials(const MuesliMAT& muesliMat, const IkarusMAT& ikarusMat) {
```
this can be called `compareIkarusAndMuesli` or similar

tarun-mitruka commented on tests/src/testmaterialswithmuesli.cpp:
```diff
@@ -0,0 +1,134 @@
+  auto stress_ikarus = ikarusMat.template stresses<strainTag>(c);
+  auto moduli_ikarus = ikarusMat.template tangentModuli<strainTag>(c);
+
+  t.check(Dune::FloatCmp::eq(energy_muesli, energy_ikarus, tol)) << testLocation();
```
use checkScalars() function from `testhelpers.hh`

henrij22 replied:
Ah and i also have to add messages

tarun-mitruka reviewed (changes requested):
Thanks Henri for this PR. Comparing it with #333, I think from this PR we not only get `LinearAnisotropic` and `LinearOrthotropic` as the new material models, but also now have an interface to work with Muesli material models, which can be used later for instance to incorporate plasticity, etc. I would also propose the following:
- Of course the container has to be fixed in the workflows.
- `LinearAnisotropic` and `LinearOrthotropic` can be tested with `planeStrain` (and `planeStress`) to see if it works correctly. This can also be investigated in a separate PR with a new issue opeend to add these tests
- Also, I think this PR can be merged after #333, as #333 also provides more rigorous tests on hyperelastic material models, to which Muesli models can also be exposed to. Also certain testhelpers are included there which can also be used here.

henrij22 committed:
trigger ci

spelling

Code review part I

license year for non-main branched files

revert test containers to official

rath3t commented on ikarus/finiteelements/mechanics/materials/muesli/mueslihelpers.hh:
```diff
@@ -0,0 +1,132 @@
+}
+
+/**
+ * \brief Converts the entries of a Eigen::Matrix to a provided muesli::tensor (symmetric 2nd order tensor).
```
All these conversions should not be needed if muesli is configured with `WITHEIGEN=TRUE`. Then istensor is basically an eigen object https://bitbucket.org/ignromero/muesli/src/27e8204971602cb042d633b8b5f87761272b10df/muesli/tensor.h?at=master#lines-55

rath3t commented on ikarus/finiteelements/mechanics/materials/muesli/mueslimaterials.hh:
```diff
@@ -0,0 +1,133 @@
+using LinearOrthotropicElasticity = muesli::elasticOrthotropicMaterial;
+
+// Alias for finite strain Muesli materials
+using StVenantKirchhoff = muesli::svkMaterial;
```
Do we really need this indirection? I would just use their namespace and we should not introduce our own `Muesli` namespace

rath3t commented on ikarus/finiteelements/mechanics/materials/muesli/mueslifinite.hh:
```diff
@@ -0,0 +1,167 @@
+  mutable itensor4 tangentModuli_{};
+
+  template <typename Derived>
+  void updateState(const Eigen::MatrixBase<Derived>& C) const {
```
A method called updateState should not be marked `const`. If the strain objects are directly Eigen types then you also do not need the evil mutable variables.

henrij22 replied:
I can of course get rid of the mutable variables, my motivation was to avoid allocating at each evaluation

henrij22 replied:
As we couldn't get Eigen to work, should we leave it like that or should I get rid of the mutable member variables?

rath3t replied:
Ok so, itensor is https://bitbucket.org/ignromero/muesli/src/c98e7d909635c1104eefd14a195970eb932302c9/muesli/Math/mtensor.h#lines-228 static sized. Therefore, i would remove the member variables and have the objects directly at call site. This also fixes the potential porlbem of race conditions sicne it is then encapsulated.

henrij22 replied:
Ah perfekt, Ill make the changes accoringly

rath3t commented on cmake/modules/AddMuesliFlags.cmake:
```diff
@@ -0,0 +1,20 @@
+
+# register all muesli related flags
+if(HAVE_MUESLI)
+  dune_register_package_flags(LIBRARIES muesli COMPILE_DEFINITIONS "ENABLE_MUESLI=1")
```
You should be able to add WITHEIGEN=1 here.

henrij22 replied:
I mark this as resolved as Muesli couldn't really be compiled with EIGEN

rath3t commented on cmake/modules/AddMuesliFlags.cmake:
```diff
@@ -0,0 +1,20 @@
+  if(HAVE_MUESLI)
+    foreach(_target ${_targets})
+      target_link_libraries(${_target} PUBLIC muesli)
+      target_compile_definitions(${_target} PUBLIC ENABLE_MUESLI=1)
```
and here.

rath3t reviewed (changes requested):
This looks quite nice already. If we merge the docker container and look into some points here, then this is mergeable.

henrij22 commented on tests/src/testlinearelasticity.cpp:
```diff
@@ -36,6 +37,12 @@ int main(int argc, char** argv) {
     return Ikarus::linearElastic(lin);
   };
+#if ENABLE_MUESLI
+  auto linearElasticFunc3D_Muesli = [](const Ikarus::YoungsModulusAndPoissonsRatio& parameter) {
```
Is this test here neccesary?

rath3t replied:
What do you mean? Is the linear elastic muesli tested anywere else?

henrij22 replied:
Yes its tested in testmaterials.cpp

henrij22 committed:
make it work again
fix all?

format ?

add tests ?

henrij22 committed:
fixes after rebase + namespace changes

henrij22 committed:
fix muesli

henrij22 committed:
fix vanihsing + correct strain Tag

henrij22 committed:
fix format

henrij22 committed:
tests

henrij22 committed:
change back to cauchy green

henrij22 committed:
save

henrij22 committed:
save

henrij22 committed:
save2

henrij22 committed:
done

henrij22 committed:
Merge branch 'main' into try/muesli

henrij22 committed:
delete bib entry

henrij22 committed:
Merge remote-tracking branch 'Origin/try/muesli' into try/muesli

rath3t commented on ikarus/finiteelements/mechanics/materials/muesli/mueslihelpers.hh:
```diff
@@ -0,0 +1,132 @@
+ * \tparam MAT the material type
+ */
+template <typename MAT>
+concept MuesliMaterialImplementation = requires { std::is_base_of_v<muesli::material, MAT>; };
```
This concept is checking the wrong thing I suppose.
requires only checks if the statements inside are valid statements. It does not check if it true or false.

https://godbolt.org/z/jPsaqGfn5

henrij22 replied:
Ah interesting

rath3t commented on ikarus/finiteelements/mechanics/materials/muesli/mueslihelpers.hh:
```diff
@@ -0,0 +1,132 @@
+ * \param it provided istensor.
+ * \param C the Eigen::Matrix that is to be converted.
+ */
+inline void toistensor(istensor& it, const Eigen::Matrix<double, 3, 3>& C) {
```
since istensor is a fixed size object i would suggest return istensor it with return instead of input-output argument by ref.

rath3t commented on ikarus/finiteelements/mechanics/materials/muesli/mueslihelpers.hh:
```diff
@@ -0,0 +1,132 @@
+ * \param it provided itensor.
+ * \param C the Eigen::Matrix that is to be converted.
+ */
+inline void toitensor(itensor& it, const Eigen::Matrix<double, 3, 3>& C) {
```
same

rath3t commented on ikarus/finiteelements/mechanics/materials/muesli/mueslihelpers.hh:
```diff
@@ -0,0 +1,132 @@
+ */
+inline auto toEigenTensor(const itensor4& it) {
+  Eigen::TensorFixedSize<double, Eigen::Sizes<3, 3, 3, 3>> moduli{};
+  moduli.setZero();
```
set zero not needed?

rath3t commented on ikarus/python/test/testmaterials.py:
```diff
@@ -365,3 +451,5 @@ def checkMaterialConstructors():
     checkVoigtTransformations()
     checkMaterialConstructors()
+
+    # instantiateMuesliMaterials()
```
Commented?

rath3t replied:
Which combination failed here?

henrij22 replied:
Ah none, this was just commented out to reduce compile time, all of these work

rath3t commented on python/ikarus/materials/__init__.py:
```diff
@@ -0,0 +1,50 @@
+from enum import Enum
+from typing import Union
+
+# The list of supported dataCollectors
```
outdated comment

rath3t commented on tests/src/testhyperelasticity.cpp:
```diff
@@ -15,8 +15,8 @@ int main(int argc, char** argv) {
-  t.subTest(testVolumetricFunctions());
-  t.subTest(recoverNeoHookeTest());
+  // t.subTest(testVolumetricFunctions());
+  // t.subTest(recoverNeoHookeTest());
```
Commented. Are these the failign testes?

henrij22 replied:
Nope, just by accident. Thanks for spotting

rath3t reviewed (commented):
See comments. looks good already. can you describe where the strange python behaviour appears? Then at least it is documented.

henrij22 committed:
Review

henrij22 committed:
fix

henrij22 committed:
fix md  (?)

Commit:
test refactor of tests

rath3t reviewed (commented):
Ok I think the problem came from the fact that for the right cauchy green tensor the strain where close to no strain(which is the identity) but for green lagrangian strain the strains are very large since there no strain is the zero matrix. But the provided strains in python are given as  right cauchy green tensor. Thus, for your fancy materials it was not possible to converge because it was so for off from zero strains. I refactored the test and now it seems to work?

Commit:
fix code style in testmaterials.py

Commit:
fix style in testmaterials.py
